### PR TITLE
DAL and BLL pt. 1

### PR DESCRIPTION
**Pull request by XeverDead**

### CODE REVIEW

File: Enigma.BLL/Encryptors/Implementations/CaesarEncryptor.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+﻿using Enigma.BLL.Encryptors.Interfaces;
+using System;
+using System.Text;
+
+namespace Enigma.BLL.Encryptors.Implementations
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Encryptors/Implementations/CaesarEncryptor.cs:5]: Why did you decide to use n-layer architecture?

File: Enigma.BLL/Services/MessageService.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        private readonly IReader<Message> _messageReader;
+        private readonly IWriter<Message> _messageWriter;
+
+        private readonly IReader<Dictionary<long, string>> _userListReader;
+
+        public MessageService(IReader<Message> messageReader, IWriter<Message> messageWriter, IReader<Dictionary<long, string>> userListReader)
+        {
+            _messageReader = messageReader ?? throw new ArgumentNullException(nameof(messageReader));
+            _messageWriter = messageWriter ?? throw new ArgumentNullException(nameof(messageWriter));
+            _userListReader = userListReader ?? throw new ArgumentNullException(nameof(userListReader));
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/MessageService.cs:22]: You are not creating an object? What is it for?

File: Enigma.BLL/Services/MessageService.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        public IEnumerable<Message> GetDialogMessages(long user1Id, long user2Id)
+        {
+            CheckUserIds(user1Id, user2Id);
+
+            var user1MessagesDirectory = Path.Combine(EnigmaSettings.MainDirectory, user1Id.ToString(), user2Id.ToString());
+            var user2MessagesDirectory = Path.Combine(EnigmaSettings.MainDirectory, user2Id.ToString(), user1Id.ToString());
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/MessageService.cs:30]: 1? 2?

File: Enigma.BLL/Encryptors/Implementations/CaesarEncryptor.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public class CaesarEncryptor : IEncryptor
+    {
+        private readonly int _key;
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Encryptors/Implementations/CaesarEncryptor.cs:9]: Wrong naming

File: Enigma.BLL/Services/MessageService.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace Enigma.BLL.Services
+{
+    public class MessageService
+    {
+        private readonly IReader<Message> _messageReader;
+        private readonly IWriter<Message> _messageWriter;
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/MessageService.cs:14]: Wrong naming

File: Enigma.BLL/Services/MessageService.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+            if (!Directory.Exists(user1MessagesDirectory))
+            {
+                Directory.CreateDirectory(user1MessagesDirectory);
+            }
+
+            if (!Directory.Exists(user2MessagesDirectory))
+            {
+                Directory.CreateDirectory(user2MessagesDirectory);
+            }
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/MessageService.cs:42]: Code repetition

File: Enigma.BLL/Services/MessageService.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            if (!_userListReader.Read(EnigmaSettings.UserListPath).ContainsKey(user1Id))
+            {
+                throw new ArgumentException("There is no user with this id", nameof(user1Id));
+            }
+
+            if (!_userListReader.Read(EnigmaSettings.UserListPath).ContainsKey(user2Id))
+            {
+                throw new ArgumentException("There is no user with this id", nameof(user2Id));
+            }
+        }
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/MessageService.cs:79]: Wrong abstraction

File: Enigma.BLL/Services/MessageService.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+        private IEnumerable<Message> GetUserToUserMessages(string directory, long senderId, long receiverId)
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/MessageService.cs:81]: Wrong abstraction

Comment (AnastasiaKorotkevich) [reply Enigma.BLL/Services/MessageService.cs:81]: Wrong naming

File: Enigma.BLL/Services/MessageService.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                return 1;
+            }
+            else if(message1.Date < message2.Date)
+            {
+                return -1;
+            }
+            else
+            {
+                return 0;
+            }
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/MessageService.cs:113]: Needs refactoring

File: Enigma.BLL/Services/UserService.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+{
+    public class UserService
+    {
+        private readonly IReader<UserCredentials> _userdataReader;
+        private readonly IWriter<UserCredentials> _userDataWriter;
+
+        private readonly IReader<Dictionary<long, string>> _userListReader;
+        private readonly IWriter<KeyValuePair<long, string>> _userListWriter;
+
+        public UserService(IReader<UserCredentials> userdataReader, IWriter<UserCredentials> userDataWriter, IReader<Dictionary<long, string>> userListReader, IWriter<KeyValuePair<long, string>> userListWriter)
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/UserService.cs:20]: Too long lines need to be broken by enter

File: Enigma.BLL/Services/UserService.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+namespace Enigma.BLL.Services
+{
+    public class UserService
+    {
+        private readonly IReader<UserCredentials> _userdataReader;
+        private readonly IWriter<UserCredentials> _userDataWriter;
+
+        private readonly IReader<Dictionary<long, string>> _userListReader;
+        private readonly IWriter<KeyValuePair<long, string>> _userListWriter;
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/UserService.cs:18]: wrong naming

File: Enigma.BLL/Services/UserService.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+            if(isDataValid)
+            {
+                var userData = _userdataReader.Read(userDataPath);
+
+                if (!((userData.Login == enteredData.Login) && (userData.Password == enteredData.Password)))
+                {
+                    isDataValid = false;
+                }
+            }
```
Comment (AnastasiaKorotkevich) [inline Enigma.BLL/Services/UserService.cs:67]: Needs refactoring

File: Enigma.Common/Models/Message.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Enigma.Common.Models
+{
+    public class Message
+    {
+        public long Id { get; set; }
+
+        public string Text { get; set; }
+
+        public long SenderId { get; set; }
```
Comment (AnastasiaKorotkevich) [inline Enigma.Common/Models/Message.cs:11]: Why is not the user model used?

File: Enigma.Common/Models/UserCredentials.cs [hunk 14]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Enigma.Common.Models
+{
+    public class UserCredentials
+    {
+        public long UserId { get; set; }
+
+        public string Login { get; set; }
+
+        public string Password { get; set; }
+    }
```
Comment (AnastasiaKorotkevich) [inline Enigma.Common/Models/UserCredentials.cs:10]: How does this model differ from the user model?

File: Enigma.Common/Settings/EnigmaSettings.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Enigma.Common.Settings
+{
+    public static class EnigmaSettings
+    {
+        public const string MainDirectory = "D:\\web";
+        public const string UserListPath = "D:\\web\\UserList.eul";
+        public const string MessageExtension = "ems";
+        public const string UserDataFileName = "data.eud";
```
Comment (AnastasiaKorotkevich) [inline Enigma.Common/Settings/EnigmaSettings.cs:8]: Needs refactoring

File: Enigma.Common/Settings/EnigmaSettings.cs [hunk 16]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Enigma.Common.Settings
+{
+    public static class EnigmaSettings
```
Comment (AnastasiaKorotkevich) [inline Enigma.Common/Settings/EnigmaSettings.cs:3]: wrong naming

File: Enigma.DAL/Readers/Implementations/MessageReader.cs [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Enigma.DAL.Readers.Implementations
+{
+    public class MessageReader : IReader<Message>
+    {
+        public Message Read(string path)
+        {
+            var message = new Message();
+
+            using (var binaryReader = new BinaryReader(File.OpenRead(path)))
```
Comment (AnastasiaKorotkevich) [inline Enigma.DAL/Readers/Implementations/MessageReader.cs:14]: What if there is no file?

Comment (XeverDead) [reply Enigma.DAL/Readers/Implementations/MessageReader.cs:14]: It'll just throw FileNotFoundException, as it should I guess.

File: Enigma.Common/Models/Message.cs (hunk 13 above)
Comment (XeverDead) [reply Enigma.Common/Models/Message.cs:11]: Most of the time messages don't have to know any user info, except sender and receiver ids.

File: Enigma.BLL/Services/MessageService.cs (hunk 2 above)
Comment (XeverDead) [reply Enigma.BLL/Services/MessageService.cs:22]: It throws ArgumentNullException if, well, any argument is null. Visual Studio suggests this way to write it, not with if.

File: Enigma.BLL/Services/MessageService.cs (hunk 8 above)
Comment (XeverDead) [reply Enigma.BLL/Services/MessageService.cs:81]: What exactly is wrong with this method? I can't quite get it.

File: Enigma.BLL/Encryptors/Implementations/CaesarEncryptor.cs (hunk 1 above)
Comment (XeverDead) [reply Enigma.BLL/Encryptors/Implementations/CaesarEncryptor.cs:5]: Basically I have three main actions in this program: read a file or write to it, do something with an information, that is stored in files, and show this information on a screen. All those actions are quite independent: part, that is doing something with information, doesn't have to now, where and how information is stored - it needs only information itself. Same is for the UI part. So using n-layer architecture is quite natural here, I guess: parts of the programm, that need each other, but don't have to know, how exactly other parts work.

**Pull request merged**